### ISSUE TITLE
Some advice on Get, Set

### ISSUE THREAD

**Issue by Varorbc** [dotnet/csharplang#1137]
# Case One:
        public double Money
        {
            get => _money;
            set => _money = value * 100;
        }
        private double _money;
# Simplify  Case One:
        public double Money
        {
            get;
~~set => value *= 100;~~

	    set => field = value * 100;
        }
~~Or
 public double Money
 {
 get => value *= 100;
 set;
 }~~
 
My point is that do not need to define _money

# Case Two:
        public string Time
        {
            get => _time.ToString("yyyyMMddHHmmss");
            set => value = null;//ignore
        }
        private DateTime _time;
# Simplify  Case Two:
        public string Time
        {
            get => value.ToString("yyyyMMddHHmmss");
~~set;//the type is DateTime~~

	    DateTime set;
        }

My point is that the Get type and the Set type can be different

Forgive me for my poor English！

**Comment by quinmars** [comment]
This is very similar to the field keyword proposal #140. The difference is that you propose to reuse the `value` keyword to get and assign the backing field. Besides that this choice is very confusing, you cannot check in the setter if the value of the property will change or not, e.g., `if (value != field) ...` vs. `if (value != value) ...`.

**Comment by MkazemAkhgary** [comment]
this proposal is very misleading because following code is already valid and it means `value=value*100`

     set => value *= 100;

the following code is also valid although its not meaningful.

    get => value.ToString("yyyyMMddHHmmss");

**Comment by Varorbc** [comment]
i see

**Comment by Varorbc** [comment]
@quinmars 
Proposal #140 is the meaning of my case one

My second case means that the input type can be different from the return type

**Comment by BobSammers** [comment]
Did you mean to write `value *= 100` in the second "Simplify Case One" example? To me that says that the value of the invisible backing field is increased every time it is read from, which is not what happens in the "Case One" it is supposed to simplify. Should it be `value * 100`, or am I missing something? I'm assuming you mean to assign the invisible backing field from `value` at the end of the setter (and presumably the getter too).

Either way, this would be a breaking change because `value` is currently a perfectly good variable name in getters and the wider scope outside the property (by "good", I obviously mean "legal"). Together with the point made by @quinmars about the ability to check existing values, I would say this means that #140 is a better option. Introducing `field` is no more or less problematic, but it's clearer in use and avoids the problem @quinmars references. I appreciate that `value *= 100` is slightly less verbose than `field = value * 100`, but I don't think the problems are worth this slim advantage.
____
I'm not sure I understand your case two. You say that getter and setter can have different types, but nothing in your suggested syntax tells the compiler that the setter should take a DateTime. I'm going to assume you mean to write something like this:
````cs
DateTime set;
````
in your replacement example.

This is quite a large conceptual change. Consider chained assignment:
````cs
a = b = c = 10;  // a, b and c are properties
````
Currently, all three properties take the value 10. This is because the assignment itself has a value (the value of `c = 10` is `10` and this is assigned to `b` and so on). This value is the value that was assigned... or is it the value of the property after assignment? This ambiguity is unimportant at the moment but would need to be explicitly determined. If you want to say that the above statement still works as it would currently, that's fine.. but then consider this:
````cs
public int number
{
  get => (int)value;
  double set;
}

int i = number = 10.23d;
Console.WriteLine(number = 12.6532);
````
The first assignment is a compiler error if an assignment is valued as the assigner but not if it's the assignee. The `WriteLine` statement will compile either way but catch people out if they get it the wrong way round.

How about:
````cs
public int number
{
  get => (int)value;
  double set;
}

number = 12.6d;
number += 12.6d;

Console.WriteLine(number);  // 24... would someone expect 25?
````
Currently, converting a double to an int needs a cast - the user is therefore explicitly aware that 0.6 is going to be lost each time. Without knowing the details of the unseen casts internal to the property, uncertainty is more likely.

None of these examples stop the proposal dead in the water, of course. Perhaps a more straightforward way of looking at this is what happens here:
````cs
x = y;
if (x == y)
  Console.WriteLine("Comparison is true");
````
The result of this comparison may or may not always be true, depending on what happens in the getters and setters, but it seems to me that introducing a construct that encourages it to be false (or even uncompilable) is not a desirable feature.

I think this will be an unexpected, possibly unwelcome and probably confusing change to the language. Similar tasks can usually be performed easily with class or extension methods. You haven't presented any use cases for this proposal. Why do you want it?

**Comment by jnm2** [comment]
I believe the `field` keyword proposed by https://github.com/dotnet/csharplang/issues/140 is the gold standard for this.

I don't like turning the setter into a value-returning expression.
I'm also confused by you setting `value = null`. What does that accomplish?

**Comment by svick** [comment]
@BobSammers 

> This value is the value that was assigned... or is it the value of the property after assignment? This ambiguity is unimportant at the moment but would need to be explicitly determined.

It's already important. The value that was assigned and the value of the property afterwards have to have the same type, but they don't have to be *the same value*.

For example, consider this code:

```c#
public static double P
{
    get => 3.14;
    set { }
}

…

double d = P = 10.23d;

Console.WriteLine(d);
Console.WriteLine(P);
```

Here, `d` is assigned the value 10.23, even though the value of `P` after the assignment is 3.14. And I suspect this is what the compiler has to do, according to the C# spec (but I haven't checked).

---

Regarding the proposal itself, I've thought it would be nice to have a property whose getter returns `IList<T>`, but whose setter accepts any `IEnumerable<T>`. But I think the confusion over what the type of the property actually is would not be worth it.

**Comment by Varorbc** [comment]
@BobSammers What you said is just what I want to express, sorry for my poor English, I did not express clearly. 
In addition, there is something I want to express in Case 2 that because sometimes the user needs to enter the DateTime type, but in the back-end processing, I hope it returns the time String type. There may be some problems in Simplify Case Two, but this is what I real need.

**Comment by BobSammers** [comment]
@svick 
> It's already important.

Good point - I should have thought about that a bit harder!
_____
@Varorbc 
In this case (and similar ones), why not just store a string (or store the DateTime and translate on use)? Your processing can be done in an extension method:
````cs
public static string MyFormat(this DateTime dt) => dt.ToString("yyyyMMddHHmmss");
...
string formattedTime = DateTime.Now.MyFormat();
````
or, if you don't like extensions, a class (or struct) initialised with a DateTime and having a member that returns a string can be produced in a few lines, or simply create a normal class method to perform the translation like the extension method code.

Any of these examples are pretty straightforward and I'd argue that they aren't just workarounds for a "missing" feature. You get:
1. Re-usability: write (and modify) in one place, use anywhere.
1. Testability: if it's written once it needs to be tested only once.
1. self-documentation - the intent of the programmer is clear at the point of use, rather than the definition of the property.

I think reuse, testability, clarity of intent and unambiguous typing more than make up for any loss of fluency.

**Comment by jnm2** [comment]
Considered in https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-04-01.md
